### DISCUSSION TITLE
Add support for base just as well as site `const { generator, site, base } = Astro`

### DISCUSSION THREAD

**Discussion by Jean-Baptiste-Lasselle** [withastro/roadmap#845] [Proposal]
### Body

# Summary

_Support for base as well as site in the global Astro_

I here propose to extend the `Astro` global, so that the following will properly compile, and both `site` and `base` set in `astro.config.mjs` are accessible in the same fashion, in `*.astro` files, like this : 

```Astro
const { generator, site, base } = Astro
```

# Background & Motivation

 

I experienced an error in my codebase, while i was trying to access the value of base, which i set from environment variable, for my astro build.

being able to set from environment variable the `base`, is something which i thinkmany dev teams would think of doing, and then they would want to have access to that value as gracefully as possible in the `*.astro` files.

see As of https://github.com/withastro/astro/issues/10007#issuecomment-1935520378 , for the full story.

# Goals

The goal is to be able to access the value of `base` set in the `astro.config.mjs`, just as well as we do with the `site` config param.

# Example 

```Astro
---
const { generator, site, base } = Astro
---

<a href={`${base}/some/link/to/another/page`}>

```

**Comment by delucis** [comment]
I agree it’s a little confusing that this is not in `Astro` like other values. You can access the value of `base` as `import.meta.env.BASE_URL` though (docs: https://docs.astro.build/en/reference/configuration-reference/#base)

**Reply by Jean-Baptiste-Lasselle** [reply]
Indeed, yet, if i set no enf variable, and just set the `base`, As of the reference you give : 

![image](https://github.com/withastro/roadmap/assets/35227860/640dcd10-0d7e-4bbc-941b-ee785a3d0cf4)


So I feel like there's something which could be even more graceful, there, wdyt?

Addendum : 

Hm actually i can access the value of `base`, using `import.meta.env.BASE_URL`, even if i didn't set any env var, it will have the value of the base defined in `astro.config.mjs`


It feels a bit weird, even though i understand the trailingSlash purpose now

![image](https://github.com/withastro/roadmap/assets/35227860/b4b607f5-7107-4373-be31-6fe524a7302d)


but still... I would expect astro to do its magic about trailing slashes without me ahving to care about it, and not to change the value i set for base, the value set is the value set, I feel i would be more readable, and healthier that a value i set, is a value i set and does not change when i refer to it... wdyt ?

**Reply by Jean-Baptiste-Lasselle** [reply]
Also, why should it be different for `site` and `base` ? I try to understand, astro is so awesome i love it and the whole work of the team and community. <3

**Reply by delucis** [reply]
Yeah, `base` and `trailingSlash` are incredibly fiddly settings, where there seems to be “no right answer” unfortunately. The reason `trailingSlash` is taken into account here is so that `import.meta.env.BASE_URL` when used alone reflects your preferred strategy, but admittedly it is a bit confusing when also using that variable to construct paths.

**Reply by Jean-Baptiste-Lasselle** [reply]
Thank you very much for the honesty of your answer, astro is awesome, I really love it, and this is not a big problem about `astro` generally speaking, even more when community users can read our messages here and are left with no misunderstanding. If only I had the room in my every day tasks, to learn what i have to learn, to contribute, I would so happily collaborate with such a great awesome, technical team. I will bear that in mind, and see asap if can can make some room to open source contribute, it is important.

**Comment by florian-lefebvre** [comment]
You can access this data through the new `astro:config` virtual modules: https://docs.astro.build/en/reference/experimental-flags/serialized-configuration/
